src/Nethermind/Nethermind.Evm.Precompiles/ModExpPrecompile.cs

```diff
@@ -146,23 +254,27 @@ public unsafe (byte[], bool) Run(ReadOnlyMemory<byte> inputData, IReleaseSpec re
         using var expInt = mpz_t.Create();
         using var powmResult = mpz_t.Create();
 
-        fixed (byte* baseData = inputData.Span.SliceWithZeroPaddingEmptyOnError(96, baseLength))
-        fixed (byte* expData = inputData.Span.SliceWithZeroPaddingEmptyOnError(96 + baseLength, expLength))
+        ReadOnlySpan<byte> baseDataSpan = inputSpan.SliceWithZeroPaddingEmptyOnError(96, (int)baseLength);
+        if (baseDataSpan.Length > 0)
         {
-            if (baseData is not null)
+            fixed (byte* baseData = &MemoryMarshal.GetReference(baseDataSpan))
                 Gmp.mpz_import(baseInt, (nuint)baseLength, 1, 1, 1, nuint.Zero, (nint)baseData);
+        }
 
-            if (expData is not null)
+        ReadOnlySpan<byte> expDataSpan = inputSpan.SliceWithZeroPaddingEmptyOnError(96 + (int)baseLength, (int)expLength);
+        if (expDataSpan.Length > 0)
+        {
+            fixed (byte* expData = &MemoryMarshal.GetReference(expDataSpan))
                 Gmp.mpz_import(expInt, (nuint)expLength, 1, 1, 1, nuint.Zero, (nint)expData);
         }
 
         Gmp.mpz_powm(powmResult, baseInt, expInt, modulusInt);
 
-        var powmResultLen = (int)(Gmp.mpz_sizeinbase(powmResult, 2) + 7) / 8;
-        var offset = modulusLength - powmResultLen;
-        byte[] result = new byte[modulusLength];
+        uint powmResultLen = (uint)(Gmp.mpz_sizeinbase(powmResult, 2) + 7) / 8;
```

Comment:
Why `uint`?

Reply:
But it still casts when dealing with `modulusLength` later.

Reply:
`uint -> nuint` (zero extend) is less expensive than `int -> nuint` (sign extend); as its essentially a null op on Intel (just use wider register `ecx => rex`; extending a singed register is more expensive as it needs to fill with `f` when top bit is set

Reply:
```csharp
// You're doing `niunt` to `uint` here:
uint powmResultLen = (uint)(Gmp.mpz_sizeinbase(powmResult, 2) + 7) / 8;
// Then creating another uint subtracting from int
uint offset = modulusLength - powmResultLen;
// Then the uint offset is added to nint ptr. Isn't it more effective to cast once to int and have it that way everywhere?
```

Reply:
mpz_sizeinbase returns a `nuint` not a signed one; will update to make more clear

Reply:
Right, so I convert to signed right away, as all others are also signed

Reply:
Lengths are all unsigned?

Reply:
GMP prefers unsigned. In this case, we have all signed values to deal with the unsigned one of GMP. So I converted it to signed right away.

Reply:
I don't understand; everything is unsigned now other than the first input to `Gmp.mpz_export` which is `nint` but is also a cast from `byte*` so 🤷‍♂️

Reply:
Also is only changing size where signed to unsigned is messy (e.g. int to ulong or nuint); so to make that less expensive you'd cast via uint e.g. `(ulong)(uint)var` as then doesn't do sign extend

Reply:
> everything is unsigned now

How's that? The unsigned value retrieved from `mpz_sizeinbase` was later used to calculate the _signed_ offset to be added to the _signed_ `ptr`. That offset is calculated by subtracting from a _signed_ modulusLength. Hence, why change things to unsigned if in the end we need it signed, and in the middle are signed values. You're changing it to unsigned just to convert it to signed later for `ptr`.

Reply:
modulusLength is unsigned; but ok https://github.com/NethermindEth/nethermind/pull/9008/commits/6fba82b0066c1f1033270c5e96d44e485007301d

Reply:
Oh, you changed it?

Reply:
That old bigint stuff looks redundant. It's used in a test to compare against the actual one. But given the number of tests we have, we can remove it to keep the code simpler. WDYT?

Reply:
Maybe. But I think it's just obsolete. We can do it in another PR, of course.

Reply:
Might still be handy as backup since we are changing the precompiles more often now